### ISSUE TITLE
KernelTransactions did't unblockNewTransactions when EmbeddedDatabase shutdown

### ISSUE THREAD

**Issue by he7g** [neo4j/neo4j#12623]
My App load data by switch two DB, shutdown old DB, and start new DB.   I found the problem is that there is a lock own the thread and did't unlock, after shutdown old DB, and the new DB create new lock to own the thread.  Detail lock is as follows:

![image](https://user-images.githubusercontent.com/22008307/99377681-75068600-2901-11eb-8b14-d201b6ecdcd0.png)

![image](https://user-images.githubusercontent.com/22008307/99377810-a41cf780-2901-11eb-8982-1085cba0ace4.png)

![image](https://user-images.githubusercontent.com/22008307/99377884-b8f98b00-2901-11eb-8f6b-706dfb0a48ee.png)

**The problem is that the old DB KernelTransactions lock is not released, and maybe cause more references are not released.**

**Comment by Roxling** [comment]
@hebelala 
Which version are you experiencing this issue in? (Also if it is Community or Enterprise edition)
I've been trying to reproduce this issue without success. Would you mind sharing the embedded code you use to stop/start the database after import?

The entire database lifecycle is recreated when the database is restarted so the `KernelTransactions` objects are never reused. Can you share a thread dump where you have a thread stuck waiting on that lock?

The only explanation I can find would be if the database failed to stop cleanly, and something suppressing the exception and continue with a start. Check the `debug.log` for any additional information  

Thanks

**Comment by he7g** [comment]
Version: Community 3.5.20 

```
package com.github.hebelala;

import org.apache.commons.io.FileUtils;
import org.neo4j.graphdb.*;
import org.neo4j.graphdb.factory.GraphDatabaseFactory;
import org.neo4j.unsafe.batchinsert.BatchInserter;
import org.neo4j.unsafe.batchinsert.BatchInserters;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Neo4jEmbeddedSwitchDemo {

	static File dataFile1 = new File("/neo4j/db1/data");
	static File dataFile2 = new File("/neo4j/db2/data");

	public static void main(String[] args) throws Exception {

		GdbWrapper gdbWrapper = null;
		while (true) {
			if (gdbWrapper == null) {
				gdbWrapper = createNew(dataFile1);
			} else {
				GdbWrapper oldGdbWrapper = gdbWrapper;
				File nextDataFile = getOtherDataFile(gdbWrapper);
				gdbWrapper = createNew(nextDataFile);
				shutdown(oldGdbWrapper);
			}
			System.out.println("sleep...");
			Thread.sleep(5000L);
		}

	}

	private static void shutdown(GdbWrapper gdbWrapper) {
		gdbWrapper.graphDatabaseService.shutdown();
		FileUtils.deleteQuietly(gdbWrapper.dataFile.getParentFile());
		gdbWrapper.nodeCache.clear();
	}

	private static GdbWrapper createNew(File dataFile) throws IOException {
		FileUtils.deleteQuietly(dataFile.getParentFile());
		importData(dataFile);
		GraphDatabaseService graphDatabaseService = new GraphDatabaseFactory().newEmbeddedDatabase(dataFile);
		createUniqueConstraint(graphDatabaseService);
		warmUpData(graphDatabaseService);
		List<String> nodeCache = buildNodeCache(graphDatabaseService);
		return new GdbWrapper(graphDatabaseService, dataFile, nodeCache);
	}

	private static void importData(File dataFile) throws IOException {
		BatchInserter batchInserter = BatchInserters.inserter(dataFile, new HashMap<>());
		try {
			Map<String, Object> properties = new HashMap<>();
			properties.put("name", "a");
			long aId = batchInserter.createNode(properties, Labels.LA);

			properties = new HashMap<>();
			properties.put("name", "b");
			long bId = batchInserter.createNode(properties, Labels.LB);

			batchInserter.createRelationship(aId, bId, RelationshipTypes.RA, new HashMap<>());
		} finally {
			batchInserter.shutdown();
		}
	}

	private static void createUniqueConstraint(GraphDatabaseService theGraphDb) {
		for (Labels label : Labels.values()) {
			String query = String.format("CREATE CONSTRAINT ON (node:%s) ASSERT node.name IS UNIQUE", label.name());
			try (Transaction transaction = theGraphDb.beginTx()) {
				theGraphDb.execute(query);
				transaction.success();
			}
		}
	}

	// https://neo4j.com/developer/kb/warm-the-cache-to-improve-performance-from-cold-start
	private static void warmUpData(GraphDatabaseService theGraphDb) {
		try (Transaction transaction = theGraphDb.beginTx()) {
			theGraphDb.execute("MATCH (n) OPTIONAL MATCH (n)-[r]->() RETURN count(n.name) + count(r)");
			transaction.success();
		}
	}

	private static List<String> buildNodeCache(GraphDatabaseService theGraphDb) {
		List<String> nodeList = new ArrayList<>();
		try (Transaction transaction = theGraphDb.beginTx()) {
			Result result = theGraphDb.execute(String.format("MATCH (n:%s) RETURN DISTINCT n", Labels.LA.name()));
			result.columnAs("n").stream().forEach(n -> {
				Node node = (Node) n;
				if (node.hasProperty("name")) {
					nodeList.add((String) node.getProperty("name"));
				}
			});
			transaction.success();
		}
		return nodeList;
	}

	private static File getOtherDataFile(GdbWrapper gdbWrapper) {
		return gdbWrapper.dataFile == dataFile1 ? dataFile2 : dataFile1;
	}

	static class GdbWrapper {
		GraphDatabaseService graphDatabaseService;
		File dataFile;
		List<String> nodeCache;

		public GdbWrapper(GraphDatabaseService graphDatabaseService, File dataFile, List<String> nodeCache) {
			this.graphDatabaseService = graphDatabaseService;
			this.dataFile = dataFile;
			this.nodeCache = nodeCache;
		}
	}

	enum Labels implements Label {
		LA, LB
	}

	enum RelationshipTypes implements RelationshipType {
		RA
	}
}

```

After print three times `sleep...`, jstack dump as follows:

```
"main" #1 prio=5 os_prio=0 tid=0x0000000003f1d000 nid=0x4070 at breakpoint[0x0000000003f0f000]
   java.lang.Thread.State: RUNNABLE
        at com.github.hebelala.Neo4jEmbeddedSwitchDemo.main(Neo4jEmbeddedSwitchDemo.java:34)

   Locked ownable synchronizers:
        - <0x00000000d70d51b0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x000000008152b660> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
```

[jstack_after_3_sleep.txt](https://github.com/neo4j/neo4j/files/5622179/jstack_after_3_sleep.txt)


More loops, more locks...

**Comment by Roxling** [comment]
@hebelala 
Hello
I've investigated this further and this is what I have found.
Not releasing locks before loosing the last reference is bad practice, but is not a problem in itself. The OpenJDK implementation is traversing the heap to find living locks, so when the finalizer is run (likely on the next GC) they should disappear. (As can be seen [here](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ConcurrentLocksPrinter.java#L70))

Are you experiencing any real issues apart from the fact that you can see them as locked on a thread dump?

I'll try to fix this is soon

**Comment by he7g** [comment]
My app process RSS is growing up all the time ( link #12627),  I found this problem within troubleshooting.  I solved this by workaround way (LifecycleListener, Reflection to release lock), but RSS problem is still existing.   So this unlock problem is not the cause.    
But my app is hard to full GC, ygc is enough because the business, so the lock references maybe exists with long time... As you said, it's bad practice.
